### DISCUSSION TITLE
LTX-2 training optimizer

### DISCUSSION THREAD

**Discussion by dorpxam** [AkaneTendo25/musubi-tuner#21] [General]
# LTX-2 OPTIMIZER (In Progress...)

## OPTIMIZER
- [x] [AdamW](#adamw-anchor-point)
- [ ] [AdamW8bit](#adamw8bit-anchor-point)
- [ ] [PagedAdamW8bit](#pagedadamw8bit-anchor-point)
- [ ] [AdaFactor](#adafactor-anchor-point)
- [ ] [Muon](#muon-anchor-point)
- [ ] [Prodigy](#prodigy-anchor-point)
- [ ] [CAME](#came-anchor-point)

Probably the Scheduler Free version of these optimizer (AdamW / Prodigy)

## TODO

Check:
https://github.com/facebookresearch/schedule_free
https://github.com/LoganBooker/prodigy-plus-schedule-free

--- 

<a name="adamw-anchor-point"></a>
## ADAMW

No installation needed, included in torch, check the [torch](https://docs.pytorch.org/docs/stable/generated/torch.optim.AdamW.html) documentation

```python
--optimizer_type AdamW
--optimizer_args 
```

<a name="adamw8bit-anchor-point"></a>
## ADAMW8BIT

Use **bitsandbytes 8-bit optimizers** internally in this fork, check the official [doc](https://github.com/bitsandbytes-foundation/bitsandbytes/blob/main/docs/source/optimizers.mdx)

```
pip install bitsandbytes
```
Usage:
```python
--optimizer_type AdamW8bit
--optimizer_args 
```

<a name="pagedadamw8bit-anchor-point"></a>
## PAGEDADAMW8BIT

Use **bitsandbytes 8-bit optimizers** internally in this fork, check the official [doc](https://github.com/bitsandbytes-foundation/bitsandbytes/blob/main/docs/source/optimizers.mdx)

```
pip install bitsandbytes
```
Usage:
```python
--optimizer_type PagedAdamW8bit
--optimizer_args 
```

<a name="adafactor-anchor-point"></a>
## ADAFACTOR

No installation needed, included in torch, check the [torch](https://docs.pytorch.org/docs/stable/generated/torch.optim.Adafactor.html) documentation

```python
--optimizer_type AdaFactor
--optimizer_args "scale_parameter=False" "relative_step=False" "weight_decay=0.01"
```

<a name="muon-anchor-point"></a>
## MUON

No installation needed, included in torch, check the [torch](https://docs.pytorch.org/docs/stable/generated/torch.optim.Muon.html) documentation (PyTorch 2.9 and later)

```python
--optimizer_type Muon
--optimizer_args "scale_parameter=False" "relative_step=False" "weight_decay=0.01"
```

<a name="prodigy-anchor-point"></a>
## PRODIGY

**Official website**: https://github.com/konstmish/prodigy

```
pip install prodigyopt
```
Usage:
```python
--optimizer_type Prodigy
--optimizer_args "d0=1e-5" "d_coef=0.4"
```

**Core Default Settings (konstmish/prodigy)**
- **Learning Rate** (`lr`): `1.0` (Recommended default; Prodigy estimates the actual learning rate).
- **Decouple Weight Decay** (`decouple`): `True` (Uses AdamW-style weight decay).
- **Weight Decay** (`weight_decay`): `0.0` (Default in raw Prodigy) or `0.01` (Often recommended in practice).
- **D Coefficient** (`d_coef`): `1.0` (Controls the learning rate estimate).
- **Betas**: `(0.9, 0.999)` or `(0.9, 0.99)`.
- **Safeguard Warmup** (`safeguard_warmup`): `True` (Recommended when using learning rate schedulers to prevent, the LR from "blowing up").
- **Use Bias Correction** (`use_bias_correction`): `True`.
- **Initial D0** (`d0`): `1e-6` (Default/Typical start). 

**Commonly Recommended Tuning (Stable Diffusion)**
While the above are the technical defaults, users often adjust them for better LoRA convergence: 
- `d_coef`: `0.5` – `0.8` (Lowering from `1.0` is common for more stable training).
- `weight_decay`: `0.01` – `0.05`.
- `decouple`: `True`. 

**Key Usage Notes**
- **LR Should be 1**: Do not set the learning rate to a small number like `1e-4`. Set it to `1` and let Prodigy find the rate.
- **Scheduler**: Often used with `constant` or `cosine` schedulers.
- **Memory**: If encountering OOM (Out of Memory) issues, you can increase `slice_p` (e.g., to `11`) to trade memory for speed.
- **Gradient Clipping**: Prodigy handles internal scaling, so it is recommended to turn off external gradient clipping. 

<a name="came-anchor-point"></a>
## CAME 

**Official website**: https://github.com/yangluo7/CAME

```
pip install came-pytorch
```
Usage:
```python
--optimizer_type came_pytorch.CAME
--optimizer_args "weight_decay=1e-2" "betas=(0.9, 0.999, 0.9999)" "eps=(1e-30, 1e-16)"
```

**Default CAME Optimizer Settings:** 
- **Learning Rate (lr):** Usually defaults to `1e-3` or `3e-4` depending on the wrapper, though `1e-3` is common for initial testing.
- **Betas $\(\beta _{1},\beta _{2}\)$:** Often `(0.9, 0.999)`, following standard Adam conventions.
- **Weight Decay:** Typically `1e-5` to `0.01` (similar to AdamW).
- **Memory Efficiency:** The core of CAME is to reduce memory from $O(nm)\$ to $O(n+m)\$ by utilizing non-negative matrix factorization (NMF) on the optimizer state. 
 
**Notes on Usage & Tuning:** 
- **Learning Rate Sensitivity:** In some implementations, the default camera optimizer learning rate (e.g. `6e-4`) might be too high, leading to blurry results in NeRF tasks; reducing it to `6e-5` may be necessary.
- **Comparison:** CAME is generally used as a memory-efficient alternative to AdamW or Adafactor, particularly for large-scale tasks like BERT pre-training.
